compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp

```diff
@@ -0,0 +1,423 @@
+        castInput.cast<BlockArgument>().getArgNumber() == inputIdx) {
+      return rewriter
+          .create(bodyOp->getLoc(), bodyOp->getName().getIdentifier(), input,
+                  castedType, bodyOp->getAttrs())
+          ->getResult(0);
+    }
+  }
+  return failure();
+}
+
+template <typename OpTy>
+static LogicalResult
+liftGenericOp(PatternRewriter &rewriter, linalg::GenericOp genericOp,
+              SmallVector<int64_t> lhsPerm, SmallVector<int64_t> rhsPerm,
+              SmallVector<int64_t> outPerm) {
+  static_assert((std::is_same<OpTy, linalg::BatchVecmatOp>::value ||
+                 std::is_same<OpTy, linalg::BatchMatvecOp>::value ||
+                 std::is_same<OpTy, linalg::BatchMatmulOp>::value) &&
+                "expected only BatchVecmatOp, BatchMatvecOp, or BatchMatmulOp");
+  LLVM_DEBUG(llvm::dbgs() << "lhsPerm: " << lhsPerm << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "rhsPerm: " << rhsPerm << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "outPerm: " << outPerm << "\n");
+  Location loc = genericOp.getLoc();
+  Value transposedLhs =
+      transposeTensor(loc, rewriter, genericOp.getInputs()[0], lhsPerm);
+  Value transposedRhs =
+      transposeTensor(loc, rewriter, genericOp.getInputs()[1], rhsPerm);
+  FailureOr<Value> extendedLhs =
+      castTensor(loc, rewriter, genericOp, 0, transposedLhs);
+  FailureOr<Value> extendedRhs =
+      castTensor(loc, rewriter, genericOp, 1, transposedRhs);
+  if (failed(extendedLhs) || failed(extendedRhs)) {
+    return failure();
+  }
+  Value genericInit = genericOp.getDpsInitOperand(0)->get();
+  SmallVector<OpFoldResult> genericMixedSizes =
+      tensor::getMixedSizes(rewriter, loc, genericInit);
+  SmallVector<OpFoldResult> batchMixedSizes =
+      applyPermutation(genericMixedSizes, invertPermutationVector(outPerm));
+  Value out = genericOp.getResults()[0];
+  auto outType = cast<RankedTensorType>(out.getType());
+  Value batchEmpty = rewriter.create<tensor::EmptyOp>(loc, batchMixedSizes,
+                                                      outType.getElementType());
+  Value zero = rewriter.create<arith::ConstantOp>(
+      loc, rewriter.getZeroAttr(outType.getElementType()));
+  Value batchInit =
+      rewriter.create<linalg::FillOp>(loc, zero, batchEmpty).getResult(0);
+  OpTy batchOp = rewriter.create<OpTy>(
+      loc, TypeRange{batchInit.getType()},
+      ValueRange{extendedLhs.value(), extendedRhs.value()},
+      ValueRange{batchInit});
+  Value newOut = transposeTensor(loc, rewriter, batchOp.getResult(0), outPerm);
+  rewriter.replaceOp(genericOp, newOut);
+  return success();
+}
+
+static LogicalResult
+raiseToBatchVecmat(PatternRewriter &rewriter, linalg::GenericOp genericOp,
+                   linalg::ContractionDimensions contractionDims) {
+  assert(contractionDims.batch.size() == 1 && contractionDims.m.size() == 0 &&
+         contractionDims.n.size() == 1 && contractionDims.k.size() == 1 &&
+         "expected batch vecmat contraction dims");
```

Comment:
Let's take a little step back and think of a slightly different interface for these helper functions that will not require duplicating this contraction-dims-matching between the caller (`matchAndRewrite`, figuring which `raise*` helper to call) and the callee (each `raise*` function validating that it was the right helper to call).

Since these helpers return `LogicalResult` already, one way to resolve this would be to replace this `assert` by just a `if(...same condition...) { return failure(); }`.  This way, the caller could just call all of the `raise*` helper functions in sequence until one succeeds. This way, these conditions would become implementation details of these helpers, letting the caller (`matchAndRewrite`) be easier to read.